src/core/dbus-execute.c

```diff
@@ -1718,20 +1730,31 @@ int bus_exec_context_set_transient_property(
                                 cpu_set_reset(affinity ? &c->cpu_set : &c->numa_policy.nodes);
                                 unit_write_settingf(u, flags, name, "%s=", name);
                         } else {
-                                _cleanup_free_ char *str = NULL;
+                                long nr_cpus;
 
-                                str = cpu_set_to_string(&set);
-                                if (!str)
-                                        return -ENOMEM;
+                                nr_cpus = sysconf(_SC_NPROCESSORS_CONF);
+                                if (nr_cpus < 0)
+                                        return -errno;
 
-                                /* We forego any optimizations here, and always create the structure using
-                                 * cpu_set_add_all(), because we don't want to care if the existing size we
-                                 * got over dbus is appropriate. */
-                                r = cpu_set_add_all(affinity ? &c->cpu_set : &c->numa_policy.nodes, &set);
-                                if (r < 0)
-                                        return r;
+                                if (affinity && CPU_ISSET_S(nr_cpus, set.allocated, set.set)) {
```

Comment:
hmm, please no, let's just add an explcit bus property for this. i.e. add CPUAffinityFromNuma as boolean bus property.

i.e. let's have a normalized API. Yes, this means that the unit files on disk and the matching dbus properties are slightly differently set up, but we do that all the time, because one is user facing and the other code facing. For user facing stuff we want the more condensed version, for code facing stuff the normalized version. 

Other cases where we do things like this are for example TasksMax=. In unit files it exists only under this name, but accepts both absolute and relative values (i.e. % values). On the wire this is mapped to two distinct properties. TasksMax is used for the absolute values and TasksMaxScale= for the relative values. 

Hence: user facing props should be close to dbus props, but should be fully normalized in their conceptual behaviour.